### ISSUE TITLE
build fails on macOS 10.12 sierra clang 9.0.0

### ISSUE THREAD

**Issue by rocksolidwebdesign** [SOCI/soci#605]
errors preventing successful build on macOS sierra occur when building from latest master, which as of this writing is: d0552e64abc2be698aaf3007a5fd5e5138f0c830

errors occur in postgresql tests and mysql tests. errors block successful library compilation and install.

errors occur both with and without `-stdlib=libc++`

build succeeds on Ubuntu 16.04 with clang++ 3.8.0 (and also g++ 5.4.0)

build succeeds on Windows 8.1 with VS2015

though I'm not sure the primary cause of the failure, I noticed while investigating some of the messages that:

comparing https://github.com/SOCI/soci/blob/master/tests/mysql/test-mysql.h#L106

to https://github.com/SOCI/soci/blob/master/tests/mysql/test-mysql.h#L116

one uses a namespace qualifier, the other does not i.e. `soci::session` vs `session`

steps to reproduce and resulting output:

```
$ cd ~/Downloads
$ git clone https://github.com/SOCI/soci.git
$ mkdir soci_build
$ cd soci_build
$ cmake -G "Unix Makefiles" -DSOCI_SHARED=ON -DSOCI_LIBDIR=lib -DSOCI_STATIC=OFF -DCMAKE_PREFIX_PATH=~/local -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/Users/me/local -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS=-stdlib=libc++ -DCMAKE_EXE_LINKER_FLAGS=-stdlib=libc++ /Users/me/Downloads/soci
-- The C compiler identification is AppleClang 9.0.0.9000037
-- The CXX compiler identification is AppleClang 9.0.0.9000037
-- Check for working C compiler: /usr/bin/clang
-- Check for working C compiler: /usr/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/clang++
-- Check for working CXX compiler: /usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for __i386__
-- Looking for __i386__ - not found
-- Looking for __x86_64__
-- Looking for __x86_64__ - found
-- Looking for __arm__
-- Looking for __arm__ - not found
-- Configuring SOCI:
-- SOCI_VERSION                             = 4.0.0
-- SOCI_ABI_VERSION                         = 4.0
-- SOCI_SHARED                              = ON
-- SOCI_STATIC                              = OFF
-- SOCI_TESTS                               = ON
-- SOCI_ASAN                                = OFF
-- SOCI_CXX_C11                             = ON
-- LIB_SUFFIX                               = 64
-- Looking for SOCI dependencies:
-- Threads:
-- CMAKE_THREAD_LIBS_INIT                   =
-- Boost:
-- Boost_RELEASE_VERSION                    = 1.63.0
-- Boost_INCLUDE_DIR                        = /Users/me/local/include
-- Boost_LIBRARIES                          = optimized /Users/me/local/lib/libboost_date_time-mt.dylib debug /Users/me/local/lib/libboost_date_time-mt-d.dylib
-- MySQL:
-- Performing Test HAVE_MYSQL_OPT_EMBEDDED_CONNECTION
-- Performing Test HAVE_MYSQL_OPT_EMBEDDED_CONNECTION - Failed
-- Found MySQL: /usr/local/mysql/include, /usr/local/mysql/lib/libmysqlclient.dylib
-- MySQL Embedded not found.
-- MYSQL_INCLUDE_DIR                        = /usr/local/mysql/include
-- MYSQL_LIBRARIES                          = /usr/local/mysql/lib/libmysqlclient.dylib
-- ODBC:
-- Looking for include files windows.h, sqlext.h
-- Looking for include files windows.h, sqlext.h - not found
-- WARNING: ODBC libraries not found, some features will be disabled.
-- Oracle:
-- ORACLE_HOME=
-- WARNING: Oracle libraries not found, some features will be disabled.
-- PostgreSQL:
-- POSTGRESQL_INCLUDE_DIRS                  = /usr/local/include
-- POSTGRESQL_LIBRARIES                     = /usr/local/lib/libpq.dylib
-- POSTGRESQL_VERSION                       = 9.5.5
-- SQLite3:
-- SQLITE3_INCLUDE_DIR                      = /usr/include
-- SQLITE3_LIBRARIES                        = /usr/lib/libsqlite3.dylib
-- Firebird:
-- SOCI_FIREBIRD_EMBEDDED                   = OFF
-- WARNING: Firebird libraries not found, some features will be disabled.
-- DB2:
-- WARNING: DB2 libraries not found, some features will be disabled.
-- Configuring SOCI core library:
-- SOCI_CORE_TARGET                         = soci_core
-- SOCI_CORE_TARGET_OUTPUT_NAME             = soci_core
-- SOCI_CORE_DEPS_LIBS                      = /usr/lib/libdl.dylib optimized /Users/me/local/lib/libboost_date_time-mt.dylib debug /Users/me/local/lib/libboost_date_time-mt-d.dylib
-- SOCI_CORE_INCLUDE_DIRS                   = /Users/me/local/include /Users/me/Downloads/soci_build /Users/me/Downloads/soci/include /Users/me/Downloads/soci_build/include /Users/me/Downloads/soci/include/private /Users/me/Downloads/soci_build/src/core
-- WITH_BOOST                               = ON
-- COMPILE_DEFINITIONS                      = CATCH_CONFIG_CPP11_NO_IS_ENUM SOCI_ABI_VERSION="4.0" HAVE_DL=1 BOOST_ALL_NO_LIB SOCI_LIB_PREFIX="libsoci_" SOCI_LIB_SUFFIX=".dylib" SOCI_DEBUG_POSTFIX=""
--
-- Configuring SOCI backend libraries:
-- Found MySQL: /usr/local/mysql/include, /usr/local/mysql/lib/libmysqlclient.dylib
-- MySQL Embedded not found.
-- ORACLE_HOME=
-- Empty - SOCI backend skeleton for new backends development
-- SOCI_EMPTY                               = ON
-- SOCI_EMPTY_TARGET                        = soci_empty
-- SOCI_EMPTY_OUTPUT_NAME                   = soci_empty
-- SOCI_EMPTY_COMPILE_DEFINITIONS           = CATCH_CONFIG_CPP11_NO_IS_ENUM SOCI_ABI_VERSION="4.0" HAVE_DL=1 BOOST_ALL_NO_LIB
-- SOCI_EMPTY_INCLUDE_DIRECTORIES           = /Users/me/local/include /Users/me/Downloads/soci_build /Users/me/Downloads/soci/include /Users/me/Downloads/soci_build/include /Users/me/Downloads/soci/include/private /Users/me/Downloads/soci/include/private
-- MySQL - SOCI backend for MySQL
-- SOCI_MYSQL                               = ON
-- SOCI_MYSQL_TARGET                        = soci_mysql
-- SOCI_MYSQL_OUTPUT_NAME                   = soci_mysql
-- SOCI_MYSQL_COMPILE_DEFINITIONS           = CATCH_CONFIG_CPP11_NO_IS_ENUM SOCI_ABI_VERSION="4.0" HAVE_DL=1 BOOST_ALL_NO_LIB
-- SOCI_MYSQL_INCLUDE_DIRECTORIES           = /Users/me/local/include /Users/me/Downloads/soci_build /Users/me/Downloads/soci/include /Users/me/Downloads/soci_build/include /Users/me/Downloads/soci/include/private /Users/me/Downloads/soci/include/private /usr/local/mysql/include
-- PostgreSQL - SOCI backend for PostgreSQL
-- SOCI_POSTGRESQL                          = ON
-- SOCI_POSTGRESQL_TARGET                   = soci_postgresql
-- SOCI_POSTGRESQL_OUTPUT_NAME              = soci_postgresql
-- SOCI_POSTGRESQL_COMPILE_DEFINITIONS      = CATCH_CONFIG_CPP11_NO_IS_ENUM SOCI_ABI_VERSION="4.0" HAVE_DL=1 BOOST_ALL_NO_LIB
-- SOCI_POSTGRESQL_INCLUDE_DIRECTORIES      = /Users/me/local/include /Users/me/Downloads/soci_build /Users/me/Downloads/soci/include /Users/me/Downloads/soci_build/include /Users/me/Downloads/soci/include/private /Users/me/Downloads/soci/include/private /usr/local/include /usr/local/include
-- SOCI_POSTGRESQL_NOPARAMS                 = OFF
-- SOCI_POSTGRESQL_NOBINDBYNAME             = OFF
-- SOCI_POSTGRESQL_NOPREPARE                = OFF
-- SOCI_POSTGRESQL_NOSINLGEROWMODE          = OFF
-- SQLite3 - SOCI backend for SQLite 3
-- SOCI_SQLITE3                             = ON
-- SOCI_SQLITE3_TARGET                      = soci_sqlite3
-- SOCI_SQLITE3_OUTPUT_NAME                 = soci_sqlite3
-- SOCI_SQLITE3_COMPILE_DEFINITIONS         = CATCH_CONFIG_CPP11_NO_IS_ENUM SOCI_ABI_VERSION="4.0" HAVE_DL=1 BOOST_ALL_NO_LIB
-- SOCI_SQLITE3_INCLUDE_DIRECTORIES         = /Users/me/local/include /Users/me/Downloads/soci_build /Users/me/Downloads/soci/include /Users/me/Downloads/soci_build/include /Users/me/Downloads/soci/include/private /Users/me/Downloads/soci/include/private /usr/include
--
-- Configuring SOCI tests:
-- SOCI_EMPTY_TEST_CONNSTR                  = dummy
-- SOCI_MYSQL_TEST_CONNSTR                  = db=soci_test
-- MS Access test disabled on non-Windows platform
-- SOCI_POSTGRESQL_TEST_CONNSTR             = dbname=soci_test
-- SOCI_SQLITE3_TEST_CONNSTR                = :memory:
--
-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   soci_core
   soci_empty
   soci_mysql
   soci_postgresql
   soci_sqlite3

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /Users/me/Downloads/soci_build
$ make -j4
Scanning dependencies of target soci_core
[  1%] Building CXX object src/core/CMakeFiles/soci_core.dir/backend-loader.cpp.o
[  5%] Building CXX object src/core/CMakeFiles/soci_core.dir/connection-parameters.cpp.o
[  4%] Building CXX object src/core/CMakeFiles/soci_core.dir/common.cpp.o
[  5%] Building CXX object src/core/CMakeFiles/soci_core.dir/blob.cpp.o
[  6%] Building CXX object src/core/CMakeFiles/soci_core.dir/connection-pool.cpp.o
[  8%] Building CXX object src/core/CMakeFiles/soci_core.dir/error.cpp.o
[  9%] Building CXX object src/core/CMakeFiles/soci_core.dir/into-type.cpp.o
[ 11%] Building CXX object src/core/CMakeFiles/soci_core.dir/once-temp-type.cpp.o
[ 12%] Building CXX object src/core/CMakeFiles/soci_core.dir/prepare-temp-type.cpp.o
[ 13%] Building CXX object src/core/CMakeFiles/soci_core.dir/procedure.cpp.o
[ 15%] Building CXX object src/core/CMakeFiles/soci_core.dir/ref-counted-prepare-info.cpp.o
[ 16%] Building CXX object src/core/CMakeFiles/soci_core.dir/ref-counted-statement.cpp.o
[ 18%] Building CXX object src/core/CMakeFiles/soci_core.dir/row.cpp.o
[ 19%] Building CXX object src/core/CMakeFiles/soci_core.dir/rowid.cpp.o
[ 20%] Building CXX object src/core/CMakeFiles/soci_core.dir/session.cpp.o
[ 22%] Building CXX object src/core/CMakeFiles/soci_core.dir/soci-simple.cpp.o
[ 23%] Building CXX object src/core/CMakeFiles/soci_core.dir/statement.cpp.o
[ 25%] Building CXX object src/core/CMakeFiles/soci_core.dir/transaction.cpp.o
[ 26%] Building CXX object src/core/CMakeFiles/soci_core.dir/use-type.cpp.o
[ 27%] Building CXX object src/core/CMakeFiles/soci_core.dir/values.cpp.o
[ 29%] Linking CXX shared library ../../lib/libsoci_core.dylib
[ 29%] Built target soci_core
Scanning dependencies of target soci_sqlite3
Scanning dependencies of target soci_mysql
Scanning dependencies of target soci_postgresql
Scanning dependencies of target soci_empty
[ 30%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/blob.cpp.o
[ 31%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/blob.cpp.o
[ 33%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/blob.cpp.o
[ 34%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/blob.cpp.o
[ 36%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/factory.cpp.o
[ 37%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/error.cpp.o
[ 38%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/common.cpp.o
[ 40%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/error.cpp.o
[ 41%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/row-id.cpp.o
[ 43%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/factory.cpp.o
[ 44%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/factory.cpp.o
[ 45%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/factory.cpp.o
[ 47%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/session.cpp.o
[ 48%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/row-id.cpp.o
[ 50%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/row-id.cpp.o
[ 51%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/row-id.cpp.o
[ 52%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/standard-into-type.cpp.o
[ 54%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/session.cpp.o
[ 55%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/session.cpp.o
[ 58%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/standard-use-type.cpp.o
[ 58%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/session.cpp.o
[ 59%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/standard-into-type.cpp.o
[ 61%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/standard-into-type.cpp.o
[ 62%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/statement.cpp.o
[ 63%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/standard-use-type.cpp.o
[ 65%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/vector-into-type.cpp.o
[ 66%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/standard-use-type.cpp.o
[ 68%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/standard-into-type.cpp.o
[ 69%] Building CXX object src/backends/empty/CMakeFiles/soci_empty.dir/vector-use-type.cpp.o
[ 70%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/statement.cpp.o
[ 72%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/standard-use-type.cpp.o
[ 73%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/statement.cpp.o
[ 75%] Linking CXX shared library ../../../lib/libsoci_empty.dylib
[ 75%] Built target soci_empty
[ 76%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/vector-into-type.cpp.o
[ 77%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/statement.cpp.o
Scanning dependencies of target soci_empty_test
[ 79%] Building CXX object tests/empty/CMakeFiles/soci_empty_test.dir/test-empty.cpp.o
[ 80%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/vector-into-type.cpp.o
[ 81%] Building CXX object src/backends/sqlite3/CMakeFiles/soci_sqlite3.dir/vector-use-type.cpp.o
[ 83%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/vector-into-type.cpp.o
[ 84%] Building CXX object src/backends/mysql/CMakeFiles/soci_mysql.dir/vector-use-type.cpp.o
[ 86%] Linking CXX shared library ../../../lib/libsoci_sqlite3.dylib
[ 86%] Built target soci_sqlite3
[ 87%] Building CXX object src/backends/postgresql/CMakeFiles/soci_postgresql.dir/vector-use-type.cpp.o
Scanning dependencies of target soci_sqlite3_test
[ 88%] Linking CXX shared library ../../../lib/libsoci_mysql.dylib
[ 90%] Building CXX object tests/sqlite3/CMakeFiles/soci_sqlite3_test.dir/test-sqlite3.cpp.o
[ 90%] Built target soci_mysql
Scanning dependencies of target soci_mysql_test
[ 91%] Building CXX object tests/mysql/CMakeFiles/soci_mysql_test.dir/test-mysql.cpp.o
[ 93%] Linking CXX shared library ../../../lib/libsoci_postgresql.dylib
[ 93%] Built target soci_postgresql
Scanning dependencies of target soci_postgresql_test
[ 94%] Building CXX object tests/postgresql/CMakeFiles/soci_postgresql_test.dir/test-postgresql.cpp.o
In file included from /Users/me/Downloads/soci/tests/mysql/test-mysql.cpp:11:
/Users/me/Downloads/soci/tests/mysql/test-mysql.h:116:34: error: reference to 'session' is ambiguous
    bool enable_std_char_padding(session& sql) const SOCI_OVERRIDE
                                 ^
/usr/include/sys/proc.h:83:8: note: candidate found by name lookup is 'session'
struct session;
       ^
/Users/me/Downloads/soci/include/soci/session.h:41:17: note: candidate found by name lookup is 'soci::session'
class SOCI_DECL session
                ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:199:9: error: reference to 'session' is ambiguous
        session sql(backEnd, connectString);
        ^
/usr/include/sys/proc.h:83:8: note: candidate found by name lookup is 'session'
struct session;
       ^
/Users/me/Downloads/soci/include/soci/session.h:41:17: note: candidate found by name lookup is 'soci::session'
class SOCI_DECL session
                ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:201:41: error: unknown type name 'sql'
        blob_table_creator tableCreator(sql);
                                        ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:201:40: error: parentheses were disambiguated as a function
      declaration [-Werror,-Wvexing-parse]
        blob_table_creator tableCreator(sql);
                                       ^~~~~
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:201:41: note: add a pair of parentheses to declare a variable
        blob_table_creator tableCreator(sql);
                                        ^
                                        (
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:205:9: error: use of undeclared identifier 'sql'
        sql << "insert into soci_test(id, img) values(7, lo_creat(-1))";
        ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:208:24: error: unknown type name 'sql'
        transaction tr(sql);
                       ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:208:23: error: parentheses were disambiguated as a function
      declaration [-Werror,-Wvexing-parse]
        transaction tr(sql);
                      ^~~~~
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:208:24: note: add a pair of parentheses to declare a variable
        transaction tr(sql);
                       ^
                       (
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:211:20: error: unknown type name 'sql'
            blob b(sql);
                   ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:211:19: error: parentheses were disambiguated as a function
      declaration [-Werror,-Wvexing-parse]
            blob b(sql);
                  ^~~~~
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:211:20: note: add a pair of parentheses to declare a variable
            blob b(sql);
                   ^
                   (
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:213:13: error: use of undeclared identifier 'sql'
            sql << "select img from soci_test where id = 7", into(b);
            ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:223:20: error: unknown type name 'sql'
            blob b(sql);
                   ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:223:19: error: parentheses were disambiguated as a function
      declaration [-Werror,-Wvexing-parse]
            blob b(sql);
                  ^~~~~
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:223:20: note: add a pair of parentheses to declare a variable
            blob b(sql);
                   ^
                   (
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:224:13: error: use of undeclared identifier 'sql'
            sql << "select img from soci_test where id = 7", into(b);
            ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:232:9: error: use of undeclared identifier 'sql'
        sql << "select img from soci_test where id = 7", into(oid);
        ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:233:9: error: use of undeclared identifier 'sql'
        sql << "select lo_unlink(" << oid << ")";
        ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:238:9: error: reference to 'session' is ambiguous
        session sql(backEnd, connectString);
        ^
/usr/include/sys/proc.h:83:8: note: candidate found by name lookup is 'session'
struct session;
       ^
/Users/me/Downloads/soci/include/soci/session.h:41:17: note: candidate found by name lookup is 'soci::session'
class SOCI_DECL session
                ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:240:41: error: unknown type name 'sql'
        blob_table_creator tableCreator(sql);
                                        ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:240:40: error: parentheses were disambiguated as a function
      declaration [-Werror,-Wvexing-parse]
        blob_table_creator tableCreator(sql);
                                       ^~~~~
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:240:41: note: add a pair of parentheses to declare a variable
        blob_table_creator tableCreator(sql);
                                        ^
                                        (
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:244:9: error: use of undeclared identifier 'sql'
        sql << "insert into soci_test(id, img) values(7, lo_creat(-1))";
        ^
/Users/me/Downloads/soci/tests/postgresql/test-postgresql.cpp:247:24: error: unknown type name 'sql'
        transaction tr(sql);
                       ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [tests/postgresql/CMakeFiles/soci_postgresql_test.dir/test-postgresql.cpp.o] Error 1
make[1]: *** [tests/postgresql/CMakeFiles/soci_postgresql_test.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 95%] Linking CXX executable ../../bin/soci_empty_test
[ 95%] Built target soci_empty_test
1 error generated.
make[2]: *** [tests/mysql/CMakeFiles/soci_mysql_test.dir/test-mysql.cpp.o] Error 1
make[1]: *** [tests/mysql/CMakeFiles/soci_mysql_test.dir/all] Error 2
[ 97%] Linking CXX executable ../../bin/soci_sqlite3_test
[ 97%] Built target soci_sqlite3_test
make: *** [all] Error 2
make -f Makefile install
[ 29%] Built target soci_core
[ 43%] Built target soci_empty
[ 58%] Built target soci_mysql
[ 73%] Built target soci_postgresql
[ 88%] Built target soci_sqlite3
[ 91%] Built target soci_empty_test
[ 93%] Building CXX object tests/mysql/CMakeFiles/soci_mysql_test.dir/test-mysql.cpp.o
In file included from /Users/me/Downloads/soci/tests/mysql/test-mysql.cpp:11:
/Users/me/Downloads/soci/tests/mysql/test-mysql.h:116:34: error: reference to 'session' is ambiguous
    bool enable_std_char_padding(session& sql) const SOCI_OVERRIDE
                                 ^
/usr/include/sys/proc.h:83:8: note: candidate found by name lookup is 'session'
struct session;
       ^
/Users/me/Downloads/soci/include/soci/session.h:41:17: note: candidate found by name lookup is 'soci::session'
class SOCI_DECL session
                ^
1 error generated.
make[2]: *** [tests/mysql/CMakeFiles/soci_mysql_test.dir/test-mysql.cpp.o] Error 1
make[1]: *** [tests/mysql/CMakeFiles/soci_mysql_test.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by mloskot** [comment]
Apparently, there is global structure named `session`:

```
In file included from /Users/me/Downloads/soci/tests/mysql/test-mysql.cpp:11:
/Users/me/Downloads/soci/tests/mysql/test-mysql.h:116:34: error: reference to 'session' is ambiguous
    bool enable_std_char_padding(session& sql) const SOCI_OVERRIDE
                                 ^
/usr/include/sys/proc.h:83:8: note: candidate found by name lookup is 'session'
struct session;
       ^
/Users/me/Downloads/soci/include/soci/session.h:41:17: note: candidate found by name lookup is 'soci::session'
class SOCI_DECL session
                ^
1 error generated.
```

Please, feel free to submit PR with fix using fully qualified name, if that solves the problem for OSX.

**Comment by rocksolidwebdesign** [comment]
this error also occurs when specifying the actual GNU based (non-apple-llvm) gcc `g++-5` from homebrew as the `-DCMAKE_CXX_COMPILER`

this means SOCI is not currently usable on macOS 10.12 sierra, regardless of toolset.

is SOCI intended to be usable on OSX? if SOCI is intended to be usable on mac then this is more than an annoyance.

are there up to date instructions for achieving a successful build on OSX?

**Comment by rocksolidwebdesign** [comment]
I would gladly submit a PR, but I am not sure what the correct fix is because I have not familiarized myself with the soci source code and I unfortunately probably don't have the time to do so.

**Comment by mloskot** [comment]
There is no strict policy on supported platforms. SOCI supports what its developers and users can test, patch and fix. Apparently, no one from the development team uses OSX, I'm afraid.

---

Regarding the issue itself, it looks like some C/C++ headers included by SOCI `session.h` also pull the POSIX `sys/proc.h` which defines `struct session` in global namespace. Obviously, this clashes with any unqualified use of `soci::session`. I'd dare to say, the problem is with C/C++ headers on OSX, which pollute global namespace with POSIX definitions (see copy of `sys/proc.h` from FreeBSD https://github.com/freebsd/freebsd/blob/e2a14c603f20af5eaf11dfafcd181d6d82269dd7/sys/sys/proc.h). 

I suspect, the same issue will happen if one compiles a sample like this:

```
#include <sys/proc.h>
struct session {};
void foo(session& s) {}
int main() {}
```

Hence, I considered it not a bug in SOCI, but an annoyance due to OSX issues.

TBH, I have no idea how to fix it apart from replacing `session` with `soci::session` wherever the ambiguity is reported on OSX.

**Comment by rocksolidwebdesign** [comment]
@mloskot 

https://github.com/SOCI/soci/pull/607

**Comment by rocksolidwebdesign** [comment]
@mloskot this PR fixes the build for me on macOS 10.12 sierra, have a look?

**Comment by mloskot** [comment]
Merged. Fixed.
